### PR TITLE
Do not show `::{{constructor}}` on tuple struct diagnostics

### LINKED ISSUE #41313: Diagnostics for tuple structs should not mention `::{{constructor}}`
To reproduce:
```rust
struct X(u32);

fn main() {
    match X(0) {
        X { data } => ()
    }
}
```
This results in:
```
error[E0026]: struct `X::{{constructor}}` does not have a field named `data`
 --> <anon>:5:13
  |
5 |         X { data } => ()
  |             ^^^^ struct `X::{{constructor}}` does not have field `data`
```
It seems rather pointless and potentially confusing to add the `::{{constructor}}` part.

### PR DESCRIPTION
**Pull request by estebank**
Fix #41313.

### CODE REVIEW

**Comment by shepmaster** [comment]
Thank you, @estebank! We'll check in every so often to make sure that @nikomatsakis or another reviewer gets back to you soon!

**Review by nikomatsakis — APPROVED** [review]

File: src/librustc/ty/item_path.rs [hunk 1]
```diff
@@ -186,6 +185,10 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
                 self.push_item_path(buffer, parent_def_id);
                 buffer.push(&data.as_interned_str());
             }
+            DefPathData::StructCtor => { // present `X` instead of `X::{{constructor}}`
```
Comment (nikomatsakis) [inline src/librustc/ty/item_path.rs]: Hmm, wait, actually -- this same code is also used for generating symbol names and the like. @michaelwoerister can you think of any complications that might arise from changing this?

Comment (michaelwoerister) [inline src/librustc/ty/item_path.rs]: This should be fine. Symbol names are fully distinguished by the hash value at the end.

**Comment by petrochenkov** [comment]
This PR is masking a deeper problem - why is a constructor DefId used in context where struct's DefId is supposed to be used? How many other contexts are affected in the same way?

I think the underlying problem need to be fixed and `::{{constructor}}` should be kept for actual constructors.
@bors r-

**Comment by Mark-Simulacrum** [comment]
Just noting that the appveyor build here failed with #40546, but not retrying because of the r- above.

**Comment by nikomatsakis** [comment]
@petrochenkov perhaps so. We do sometimes use the constructor def-id for tuple structs, I think? But it's been a while since I looked at this code. Worth investigating. 

@estebank is it easy for you to get a `RUST_BACKTRACE=1` where the error occurs? (You may find `-Ztreat-err-as-bug` helpful.) That would help pinpoint where the error is being reported.

**Comment by estebank** [comment]
@nikomatsakis @petrochenkov I believe this is caused because of [these lines in `collect.rs`](https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/collect.rs#L663-L674):

```rust
        ItemStruct(ref def, _) => {
            // Use separate constructor id for unit/tuple structs and reuse did for braced structs.
            let ctor_id = if !def.is_struct() {
                Some(tcx.hir.local_def_id(def.id()))
            } else {
                None
            };
            (AdtKind::Struct, vec![
                convert_struct_variant(tcx, ctor_id.unwrap_or(def_id), item.name,
                                       ty::VariantDiscr::Relative(0), def)
            ])
        }
```

I'm intrigued by the comment as it says what it's doing, but it doesn't state _why_. It was originally incorporated in 2cdd9f1c97652ee799f6bb0af7c063115ff368ea, I need to further read that commit to understand why.

**Comment by petrochenkov** [comment]
@estebank 
I recall this now.
The situation with constructor `DefId`s is a bit of mess, and quite large refactoring is required to fix it. Structs and their constructors have different DefIds while variants have the same DefId for both variants themselves and constructors. Variants historically keep constructor ids if constructor exists and struct ids otherwise.
 See https://github.com/rust-lang/rust/pull/28816 (including "outdated" hidden comments) and https://github.com/rust-lang/rust/pull/36814#discussion_r81044733 for some background.

>It was originally incorporated in 2cdd9f1

That commit only adds the comment, the `DefId`-assigning logic was pre-existing.

In the meantime, local fix can make diagnostics better:
```diff
diff --git a/src/librustc_typeck/check/_match.rs b/src/librustc_typeck/check/_match.rs
index 1086773041..450b0f95fc 100644
--- a/src/librustc_typeck/check/_match.rs
+++ b/src/librustc_typeck/check/_match.rs
@@ -660,8 +660,14 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                                etc: bool) {
         let tcx = self.tcx;

+        let mut diagn_did = variant.did;
         let (substs, kind_name) = match adt_ty.sty {
-            ty::TyAdt(adt, substs) => (substs, adt.variant_descr()),
+            ty::TyAdt(adt, substs) => {
+                if !adt.is_enum() {
+                    diagn_did = tcx.parent_def_id(variant.did).expect("no struct def id");
+                }
+                (substs, adt.variant_descr())
+            }
             _ => span_bug!(span, "struct pattern is not an ADT")
         };

@@ -700,12 +706,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             struct_span_err!(tcx.sess, span, E0026,
                                              "{} `{}` does not have a field named `{}`",
                                              kind_name,
-                                             tcx.item_path_str(variant.did),
+                                             tcx.item_path_str(diagn_did),
                                              field.name)
                                 .span_label(span,
                                             &format!("{} `{}` does not have field `{}`",
                                                      kind_name,
-                                                     tcx.item_path_str(variant.did),
+                                                     tcx.item_path_str(diagn_did),
                                                      field.name))
                                 .emit();
```

**Comment by petrochenkov** [comment]
Actually, doesn't matter.
@bors r=michaelwoerister

**Pull request merged**